lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp

```diff
@@ -0,0 +1,254 @@
+    shouldBeRemoved.push_back(it->second);
+    return src;
+  }
+
+  auto type = getTypeFromFieldRef(fieldRef);
+  auto builder = OpBuilder::atBlockEnd(moduleOp.getBody());
+
+  auto getMergedValue = [&](auto aggregateType) {
+    SmallVector<Value> operands;
+
+    // This flag tracks whether we can use an aggregate value as merged value.
+    bool canUseSourceAggregate = true;
+
+    // The value which might be used as a merged value.
+    Value sourceAggregate;
+
+    auto checkSourceAggregate = [&](auto subelement, unsigned destIndex,
+                                    unsigned sourceIndex) {
+      // In the first iteration, register the source aggregate value.
+      if (destIndex == 0) {
+        if (subelement.input().getType() == type)
+          sourceAggregate = subelement.input();
+        else {
+          // If types are not same, it is not possible to use it.
+          canUseSourceAggregate = false;
+        }
+      }
+
+      // Check that input is the same as `sourceAggregate` and indexes match.
+      canUseSourceAggregate &=
+          subelement.input() == sourceAggregate && destIndex == sourceIndex;
+    };
+
+    for (auto idx : llvm::seq(0u, (unsigned)aggregateType.getNumElements())) {
+      auto offset = aggregateType.getFieldID(idx);
+      auto value = getMergedConnectedValue(fieldRef.getSubField(offset));
+      if (!value)
+        return Value();
+
+      operands.push_back(value);
+
+      // From here, check whether the value is derived from the same aggregate
+      // value.
+
+      // If canUseSourceAggregate is aldready false, abort.
+      if (!canUseSourceAggregate)
+        continue;
+
+      // If the value is an argument, it is not derived from an aggregate value.
+      if (!value.getDefiningOp()) {
+        canUseSourceAggregate = false;
+        continue;
+      }
+
+      TypeSwitch<Operation *>(value.getDefiningOp())
+          .template Case<SubfieldOp>([&](SubfieldOp subfield) {
+            checkSourceAggregate(subfield, idx, subfield.fieldIndex());
+          })
+          .template Case<SubindexOp>([&](SubindexOp subindex) {
+            checkSourceAggregate(subindex, idx, subindex.index());
+          })
+          .Default([&](auto) { canUseSourceAggregate = false; });
+    }
+
+    // If it is fine to use `sourceAggregate` as a merged value, we just
+    // return it.
+    if (canUseSourceAggregate) {
+      LLVM_DEBUG(llvm::dbgs() << "Success to merge " << fieldRef.getValue()
+                              << " ,fieldID= " << fieldRef.getFieldID() << "to"
+                              << sourceAggregate << "\n";);
+      return sourceAggregate;
+    }
+
+    // Otherwise, we concat all values and cast them into the aggregate type.
+    Value accumulate;
+    changed = true;
+    for (auto value : operands) {
+      value = builder.createOrFold<BitCastOp>(
+          value.getLoc(),
+          UIntType::get(value.getContext(),
+                        *firrtl::getBitWidth(
```

Comment:
Did I miss the check that all the components have fixed width and are integer-like?  (at least are legal to concat?)
What about flip types in the aggregate?

Reply:
I had implicit assumptions that inputs came through inferWidth and LowerTypes with current aggregate preservation.
They guarantee that all preserved aggregate types don't contain analog-type or flip type. I agree that I should check these assumption explicitly so I'll add.

Reply:
Added checks at begging of `peelConnection`